luarules/gadgets/unit_carrier_spawner.lua

```diff
@@ -350,14 +351,79 @@ local function UnDockUnit(unitID, subUnitID)
 		carrierMetaList[unitID].subUnitsList[subUnitID].docked = false
 		carrierMetaList[unitID].activeDocking = false
 		carrierMetaList[unitID].subUnitsList[subUnitID].activeDocking = false
-		unitUndocked = Spring.CallCOBScript(subUnitID, "Undocked", 0, carrierMetaList[unitID].cobundockparam, carrierMetaList[unitID].subUnitsList[subUnitID].dockingPiece)
+		carrierMetaList[unitID].subUnitsList[subUnitID].activeUndockSequence = false
+		Spring.CallCOBScript(subUnitID, "Undocked", 0, carrierMetaList[unitID].cobundockparam, carrierMetaList[unitID].subUnitsList[subUnitID].dockingPiece)
 		if carrierMetaList[unitID].dockArmor then
 			spSetUnitArmored(subUnitID, false, 1)
 		end
 	end
 end
 
 
+local function UnDockSequence(unitID, subUnitID) -- Initiates an undock sequence in COB
+	if not carrierMetaList[unitID] then
+		return
+	elseif not carrierMetaList[unitID].subUnitsList[subUnitID] then
+		return
+	elseif carrierMetaList[unitID].subUnitsList[subUnitID].docked == true and not carrierMetaList[unitID].subUnitsList[subUnitID].stayDocked then
+		if carrierMetaList[unitID].subUnitsList[subUnitID].activeUndockSequence == true then
+			return
+		elseif carrierMetaList[unitID].droneundocksequence then
+			Spring.CallCOBScript(unitID, "DroneUndockSequence", 0, subUnitID, carrierMetaList[unitID].subUnitsList[subUnitID].dockingPiece)
+			Spring.CallCOBScript(subUnitID, "DroneUndockSequence", 0, carrierMetaList[unitID].cobundockparam, carrierMetaList[unitID].subUnitsList[subUnitID].dockingPiece)
```

Comment:
This assumes a unit can never be both a drone _and_ a carrier, which seems a reasonable design choice.